### PR TITLE
Fix AD0001 on S1186: NullReferenceException for LocalFunctionStatements

### LINKED ISSUE #7050: Fix AD0001 on S1186: NullReferenceException for top-level methods
### Description

When applying `EmptyMethodCodeFix` to an empty top-level method (available from C# 9.0) the analyzer throws an exception (AD0001):

```
Stack Trace: 
EmptyMethodCodeFix.RegisterCodeFixesAsync(SyntaxNode root, CodeFixContext context) line 51
SonarCodeFix.RegisterCodeFixesAsync(CodeFixContext context) line 45
--- End of inner exception stack trace ---
Task.ThrowIfExceptional(Boolean includeTaskCanceledExceptions)
Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
Task.Wait()
CodeFixVerifier.ActionToApply(CodeFixProvider codeFix, Document document, Diagnostic diagnostic) line 94
<>c__DisplayClass6_0.<VerifyWhileDocumentChanges>b__1(Diagnostic x) line 56
SelectManySingleSelectorIterator`2.MoveNext()
Enumerable.TryGetFirst[TSource](IEnumerable`1 source, Boolean& found)
Enumerable.FirstOrDefault[TSource](IEnumerable`1 source)
CodeFixVerifier.VerifyWhileDocumentChanges(ParseOptions parseOptions, String pathToExpected) line 54
Verifier.VerifyCodeFix() line 117
VerifierBuilderExtensions.VerifyCodeFix(VerifierBuilder builder) line 161
```

### Repro steps

In Visual Studio put the cursor on `EmptyTopLevelMethod` in the snippet below, press `ctrl + dot` and see the analyzer except.
```cs
void EmptyTopLevelMethod() { }
```

### Note

The `RegisterCodeFixesAsync` throws a `NullReferenceException`, after it unsuccessfully attempts to find the method declaration node:
`syntaxNode.FirstAncestorOrSelf<MethodDeclarationSyntax>();`
A top-level method is always considered a local function (`LocalFunctionStatementSyntax`) rather than a `MethodDeclarationSyntax`.
This can be reproduced in a unit test by adding a code fix test for an empty top-level method.

Another problem - caused by the same issue - is when a local function is inside another method:

```cs
public class MyClass
{
    public void Method()
    {
        Console.WriteLine();

        void LocalFunction()
        {
        }
    }
}
```

If we apply the Throw fix for `LocalFunction`, then it will remove the code of the outer method (as well as the local function) and replace it with a throw statement:

```cs
public class MyClass
{
    public void Method()
    {
        throw new NotSupportedException();
    }
}
```

Summa summarum, additional test cases are needed for local functions in general, not just for top-level methods when code fixes are involved.

### Related information

* C#/VB.NET Plugins version
  * SonarLint 6.14.1.66430
* Visual Studio version: 
  * Version 17.5.4
* Operating System: 
  * Windows 10

### PR DESCRIPTION
**Pull request by csaba-sagi-sonarsource**
Fixes #7050

### CODE REVIEW

**Comment by csaba-sagi-sonarsource** [comment]
It is reviewable, I converted it to draft to make sure it does not get merged into 9.1

File: analyzers/src/SonarAnalyzer.CSharp/Rules/EmptyMethodCodeFix.cs [hunk 1]
```diff
@@ -111,24 +106,19 @@ private static async Task RegisterCodeFixesForMethodsAsync(CodeFixContext contex
...
-            return !semanticModel.LookupNamespacesAndTypes(method.Body.CloseBraceToken.SpanStart)
+        private static bool NamespaceNeedsToBeAdded(BlockSyntax methodBody, SemanticModel semanticModel) =>
+            !semanticModel.LookupNamespacesAndTypes(methodBody.CloseBraceToken.SpanStart)
                 .OfType<INamedTypeSymbol>()
-                .Any(nt =>
-                    nt.IsType &&
-                    nt.Name == LiteralNotSupportedException &&
-                    nt.ContainingNamespace.Name == LiteralSystem);
-        }
+                .Any(nt => nt.IsType
```
Comment (costin-zaharia-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/EmptyMethodCodeFix.cs:120]: you can also rename `nt` to `symbol` since we don't use abbreviations.

Comment (csaba-sagi-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/EmptyMethodCodeFix.cs:120]: Thanks for pointing this out! I missed it completely.
According to our coding style: Single variable lambdas should use x as the variable name (based on lambda calculus λx)
So, I went for x.

**Review by costin-zaharia-sonarsource — APPROVED** [review]
LGTM!
